documentation/docs/explanations/deadlock-analysis.mdx

```diff
@@ -0,0 +1,179 @@
+can access to a shared memory.
+
+#### Model of function and locks
+Here how we represent a function and locks.
+
+![](./assets/function_model.drawio.png)
+
+- $1$, $2$, $3$ represent locks that protect 3 different and distinct shared memory locations
+- $f_1$ is a function that sequentially takes hold of lock $1$ then $2$ then $3$
+- $f_1$ releases the locks $1$, $2$, $3$ at the end of its computation 
+
+Here we can see that lock $2$ depends on lock $1$ because in $f_1$ you can't take lock $2$ before taking lock $1$. In the same way lock $3$ depends on locks $1$ and $2$.
+
+#### Example: Deadlock with 2 functions
+Here is an example of the most simple possible deadlock.
+
+![](assets/deadlock_2_functions.drawio.png)
+
+- the function $f_1$ wants to take the locks for shared memory $1$ then $2$
+- the function $f_2$ wants to take the locks for shared memory $2$ then $1$
+- on the right you can see that there is a circular dependency of locks between $f_1$ and $f_2$ (you can do a circle by following the arrows of $f_1$ and $f_2$). 
+
+##### Analysis
+
+The graph at the right represents the merged dependency graph of functions $f_1$ and $f_2$. It shows all the dependencies of the locks in $f_1$ and $f_2$.
+The presence of a circular dependency on the merged graph indicates the possibility of a deadlock when executing $f_1$ and $f_2$ concurrently.
+
+##### Example run resulting in a deadlock: 
+1. $f_1$ starts and takes the lock $1$
+2. $f_2$ starts and takes the lock $2$
+3. $f_1$ is stuck because $f_2$ is holding the lock $2$ that it requires
+4. $f_2$ is stuck because $f_2$ is holding the lock $1$ that it requires
+
+
+
+#### Example: Deadlock with 3 functions
+![](assets/deadlock_3_functions.drawio.png)
+
+As the previous example there is a circular dependency of locks 
+between $f_1$, $f_2$ and $f_3$. 
+Example run resulting in a deadlock: 
+1. $f_1$ starts and takes the lock $1$
+2. $f_2$ starts and takes the lock $3$
+3. $f_3$ starts and takes the lock $2$
+4. $f_1$ is stuck because $f_3$ is holding the lock $2$ that it requires
+5. $f_2$ is stuck because $f_1$ is holding the lock $1$ that it requires
+6. $f_3$ is stuck because $f_2$ is holding the lock $3$ that it requires
+
+### Proof backing "_no circular dependency equals to no deadlock_"
+This is a proof on an example for simplicity. The reasoning used in the next part can be applied to any dependency graph of locks to prove the absence of deadlock when there are no circular dependency.
+
+#### Example: no deadlock
+![](assets/no_deadlock.drawio.png)
+
+There is no circular dependency in the right-side representation.
+We will try to argue that there is indeed no possible deadlock in this configuration.
+The main idea linking absence of deadlock with absence of circular dependency is that any thread will always eventually get the lock it's waiting for. Thus it will eventually terminate its computation thus releasing the locks it acquired.
```

Comment:
```suggestion
#### Example: no deadlock

![No deadlock](assets/no_deadlock.drawio.png)

There is no circular dependency in the right-side representation.
We will try to argue that there is indeed no possible deadlock in this configuration.
The main idea linking absence of deadlock with absence of circular dependency is that any thread will always eventually get the lock it's waiting for. So, it will eventually terminate its computation thus releasing the locks it acquired.
```